### PR TITLE
Add Proxy-Status HTTP response header field

### PR DESCRIPTION
**Pull request by ambuc**
Signed-off-by: James Buckland jbuckland@google.com

Commit Message: Add Proxy-Status HTTP response header field.
Additional Description: This PR adds the new "Proxy-Status" HTTP response header field, as proposed in #17439. See [the RFC draft](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-proxy-status-08) for much more detail.

This new proposal "defines the Proxy-Status HTTP field to convey the details of intermediary response handling, including generated errors." Proxy-Status is meant to be generally applicable to many kinds of proxies. This response header has an error field and a more general details field.

Envoy should generate and attach this new header to traffic.

Implementing this new web standard will help people who use Envoy (and people who use tools which use Envoy) better understand the state of each intermediate proxy along a connection.

The RFC is expected to be approved with minimal content changes. We should keep this new header off by default, and make clear in the documentation that it is experimental and subject to change, until the RFC is finalized and approved.

(Author's note: I last worked on this in PR https://github.com/envoyproxy/envoy/pull/17638, but then I waited too long and the PR decayed a bit too much with respect to HEAD. It was much simpler to recreate the changes new atop HEAD. Sorry for the duplicate work.)

Risk Level: Low
Testing: Unit tests for HCM in conn_manager_impl_test.cc and utils unit test in stream_info/utility_test.cc.
Docs Changes: n/a, except for self-documenting (hopefully) proto config changes.
Release Notes: n/a
Platform Specific Features: n/a
API Considerations: n/a, unsure how much to document -- this won't switch anything on by default, but users can enable a new HTTP response header field.

### CODE REVIEW

File: source/common/stream_info/utility.cc
Comment (adisuissa) [inline source/common/stream_info/utility.cc:335]: Is there a map somewhere between a response-flag and the corresponding `ProxyStatusError` enum value?
If so, can you add a link in the comment above.

File: test/common/http/conn_manager_impl_test.cc
Comment (adisuissa) [inline test/common/http/conn_manager_impl_test.cc:3727]: Overriding the test framework TearDown() method should remove these calls from each test.

**Review by adisuissa — COMMENTED** [review]
/lgtm api
Code LGTM, left minor comments.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (ambuc) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:309]: Done.

File: test/common/http/conn_manager_impl_test.cc
Comment (ambuc) [reply test/common/http/conn_manager_impl_test.cc]: Done.

File: source/common/stream_info/utility.cc
Comment (ambuc) [reply source/common/stream_info/utility.cc:335]: There isn't a canonical mapping besides what I've sketched here. I figure we can continue to iterate on these values over time, since the mapping is inherently lossy, and the RFC is still being finalized. I also didn't think it would be useful to encode this mapping in a const CONSTRUCT_ON_FIRST_USE map and iterate through the ResponseFlag values, since there's no easy ordered iterator through those values.

**Comment by jmarantz** [comment]
Yan do you want to do a maintainer pass and then assign to a non-google senior maintainer?

**Comment by jmarantz** [comment]
I poked around the coverage failure and I think it's actually due to this PR: `Code coverage for source/common/stream_info is lower than limit of 96.6 (78.7)`

and the artifacts (hard to fine but they are findable): https://storage.googleapis.com/envoy-pr/9ac6f16/coverage/source/common/stream_info/utility.cc.gcov.html

Based on the content of this PR I'd say we either added some new code that is not well unit-tested, or we accidentally removed some test coverage for previously existing but no-longer-interesting code. James is either of those possible?

**Comment by ambuc** [comment]
We definitely added some new code (per https://storage.googleapis.com/envoy-pr/9ac6f16/coverage/source/common/stream_info/utility.cc.gcov.html, thanks very much for this link) which is under-unit-tested. I'll add unit tests to bring this back to 100%.

File: source/common/http/conn_manager_impl.cc
Comment (yanavlasov) [inline source/common/http/conn_manager_impl.cc:1472]: It makes sense to move this into a dedicated method and then call it from the `ConnectionManagerUtility::mutateResponseHeaders` to keep all the response header mutations in one place.

File: source/common/stream_info/utility.cc
Comment (yanavlasov) [inline source/common/stream_info/utility.cc:194]: nit: maybe make this method name more descriptive as it is not clear what exactly is converted to string.

**Review by yanavlasov — CHANGES_REQUESTED** [review]

File: source/common/stream_info/utility.cc
Comment (ambuc) [reply source/common/stream_info/utility.cc]: Done.

File: source/common/http/conn_manager_impl.cc
Comment (ambuc) [reply source/common/http/conn_manager_impl.cc]: Done. I made a new method named ConnectionManagerUtility::setProxyStatusHeader., then called that method from ConnectionManagerUtility::mutateResponseHeaders. This necessitated adding two new arguments to ConnectionManagerUtility::mutateResponseHeaders and propagating that change to all callsites and test callsites, but it seemed like the right move -- understood that it's best to keep all response header mutation together.

**Comment by alyssawilk** [comment]
cc @adisuissa and @yanavlasov ping!

**Review by adisuissa — CHANGES_REQUESTED** [review]
Code LGTM, thanks!
Please add an entry in the release notes (new feature).

/wait

File: docs/root/version_history/current.rst
Comment (adisuissa) [inline docs/root/version_history/current.rst:29]: See result in: https://storage.googleapis.com/envoy-pr/4f9da61/docs/version_history/current.html
RST uses a different external link style (https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#embedded-uris-and-aliases).

Should be
```suggestion
* http: added support for :ref:`proxy_status_config <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.proxy_status_config>` for configuring `Proxy-Status <https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-proxy-status-08>`_ HTTP response header fields.
```

**Review by adisuissa — CHANGES_REQUESTED** [review]
One minor update to release note.
/wait

**Review by adisuissa — APPROVED** [review]
/lgtm api
LGTM code changes

File: source/common/http/conn_manager_config.h
Comment (yanavlasov) [inline source/common/http/conn_manager_config.h:507]: nit: the only small suggestion I have is to add `nameForProxyStatus()` method to the config and set it to either node ID (you can get it from the `context` constructor parameter) or the envoy literal. In this case you do not need to plumb the node id into the `mutateResponseHeaders` and can just get it from the config.

This is not a big deal, but might be helpful if we later need to be able to specify proxy name in the proxy status config.

File: source/common/stream_info/utility.h
Comment (lizan) [inline source/common/stream_info/utility.h:138]: C++ nit: avoid top-level const in declarations since they does nothing (throughout the file)

https://clang.llvm.org/extra/clang-tidy/checks/readability-avoid-const-params-in-decls.html

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (lizan) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:323]: I don't think `enum` is enough for future reservation here. Should be a message with `oneof`, some downstream build might want other literals.

File: source/common/http/conn_manager_utility.cc
Comment (lizan) [inline source/common/http/conn_manager_utility.cc:453]: Instead of passing `node_id` around, why not passing cached proxy name? If we have more options for proxy name this list doesn't grow then.

**Review by lizan — CHANGES_REQUESTED** [review]

File: source/common/stream_info/utility.h
Comment (ambuc) [reply source/common/stream_info/utility.h:138]: Done.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (ambuc) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto]: Thanks, this is a great idea. Done.

File: source/common/http/conn_manager_utility.cc
Comment (ambuc) [reply source/common/http/conn_manager_utility.cc:453]: Could you point me to the cached proxy name value? Sorry, I was unable to find it.

Comment (lizan) [reply source/common/http/conn_manager_utility.cc:453]: I meant you can pre-calculate it (and store in HCM) instead of passing `node_id` around.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (lizan) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:350]: I don't think you even need the enum, but just node_id option? since this `oneof` is already extendable, and you can default to `envoy` if none of them are set.

Comment (ambuc) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto]: Thanks, this is a really good idea.

File: source/common/http/conn_manager_utility.cc
Comment (ambuc) [reply source/common/http/conn_manager_utility.cc]: Thanks, this is a really excellent idea. Implemented in 071c406, ptal.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (lizan) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:343]: comment on each field.

`literal_proxy_name` should default to [`server_name`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#envoy-v3-api-field-extensions-filters-network-http-connection-manager-v3-httpconnectionmanager-server-name), whose default is `envoy`.

**Review by lizan — COMMENTED** [review]
Thanks, one last comment.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
Comment (ambuc) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto]: Thanks, done.

**Review by lizan — APPROVED** [review]

**Comment by lizan** [comment]
LGTM, defer to @yanavlasov to merge.

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
